ark-core/src/arknote.rs

```diff
@@ -0,0 +1,826 @@
+use crate::Error;
+use crate::VirtualUtxoScript;
+use bitcoin::hashes::sha256;
+use bitcoin::hashes::Hash;
+use bitcoin::key::Secp256k1;
+use bitcoin::opcodes::all::*;
+use bitcoin::Amount;
+use bitcoin::OutPoint;
+use bitcoin::ScriptBuf;
+use bitcoin::TxOut;
+use serde::Deserialize;
+use serde::Serialize;
+use std::fmt;
+
+/// Default human-readable prefix for ArkNote string encoding
+pub const DEFAULT_HRP: &str = "arknote";
+
+/// Length of the preimage in bytes
+pub const PREIMAGE_LENGTH: usize = 32;
+
+/// Length of the value field in bytes
+pub const VALUE_LENGTH: usize = 4;
+
+/// Total length of an encoded ArkNote
+pub const ARKNOTE_LENGTH: usize = PREIMAGE_LENGTH + VALUE_LENGTH;
+
+/// Fake outpoint index used for ArkNotes
+pub const FAKE_OUTPOINT_INDEX: u32 = 0;
+
+/// Status of a coin/VTXO
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Status {
+    pub confirmed: bool,
+}
+
+/// Extended coin trait that ArkNote implements
+pub trait ExtendedCoin {
+    fn txid(&self) -> &str;
+    fn vout(&self) -> u32;
+    fn value(&self) -> Amount;
+    fn status(&self) -> &Status;
+    fn extra_witness(&self) -> Option<&[Vec<u8>]>;
+    fn tap_tree(&self) -> Vec<String>;
+    fn forfeit_tap_leaf_script(&self) -> &ScriptBuf;
+    fn intent_tap_leaf_script(&self) -> &ScriptBuf;
+}
+
+impl fmt::Display for ArkNote {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let encoded = self.encode();
+        let value = format!("{}{}", self.hrp, bs58::encode(encoded).into_string());
+        write!(f, "{value}")
+    }
+}
+
+/// ArkNote is a fake VTXO coin that can be spent by revealing the preimage
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ArkNote {
+    preimage: [u8; PREIMAGE_LENGTH],
+    value: Amount,
+    hrp: String,
+    // Computed fields
+    txid: String,
+    vtxo_script: VirtualUtxoScript,
+    tap_tree_bytes: Vec<String>, // Cache for tap_tree() method
+    status: Status,
+    extra_witness: Vec<Vec<u8>>,
+}
+
+impl ArkNote {
+    /// Create a new ArkNote with the given preimage and value
+    pub fn new(preimage: [u8; PREIMAGE_LENGTH], value: Amount) -> Self {
+        Self::new_with_hrp(preimage, value, DEFAULT_HRP.to_string())
+    }
```

Comment:
I guess this is supposed to be a private method for testing, given that the `CreateNote` API suggests that notes are generated on the server.

Reply:
In general, let's make sure we only mark as public whatever is actually meant to be public.